sklearnex/cluster/k_means.py

```diff
@@ -1,17 +1,370 @@
+
+            self._onedal_estimator = onedal_KMeans(**onedal_params)
+
+        def _onedal_fit_supported(self, method_name, X, y=None, sample_weight=None):
+            assert method_name == "fit"
+
+            class_name = self.__class__.__name__
+            patching_status = PatchingConditionsChain(f"sklearn.cluster.{class_name}.fit")
+
+            sample_count = _num_samples(X)
+            self._algorithm = self.algorithm
+            supported_algs = ["auto", "full", "lloyd", "elkan"]
+            correct_count = self.n_clusters < sample_count
+            is_data_supported = (
+                _is_csr(X) and daal_check_version((2024, "P", 600))
+            ) or not issparse(X)
+            sample_weight = _check_sample_weight(
+                sample_weight, X, dtype=X.dtype if hasattr(X, "dtype") else None
+            )
+
+            patching_status.and_conditions(
+                [
+                    (
+                        self.algorithm in supported_algs,
+                        "Only lloyd algorithm is supported, elkan is computed using lloyd",
+                    ),
+                    (correct_count, "n_clusters is smaller than number of samples"),
+                    (
+                        np.allclose(sample_weight, np.ones_like(sample_weight)),
+                        "Sample weights are not ones.",
+                    ),
+                    (
+                        is_data_supported,
+                        "Supported data formats: Dense, CSR (oneDAL version >= 2024.6.0).",
+                    ),
+                ]
+            )
+
+            return patching_status
+
+        def fit(self, X, y=None, sample_weight=None):
+            if sklearn_check_version("1.0"):
+                self._check_feature_names(X, reset=True)
+            if sklearn_check_version("1.2"):
+                self._validate_params()
+
+            dispatch(
+                self,
+                "fit",
+                {
+                    "onedal": self.__class__._onedal_fit,
+                    "sklearn": sklearn_KMeans.fit,
+                },
+                X,
+                y,
+                sample_weight,
+            )
+
+            return self
+
+        def _onedal_fit(self, X, _, sample_weight, queue=None):
+            X = self._validate_data(
+                X,
+                accept_sparse="csr",
+                dtype=[np.float64, np.float32],
+            )
+
+            if sklearn_check_version("1.2"):
+                self._check_params_vs_input(X)
+            else:
+                self._check_params(X)
+
+            self._n_features_out = self.n_clusters
+            self._n_threads = _openmp_effective_n_threads()
+
+            self._initialize_onedal_estimator()
+            self._onedal_estimator.fit(X, queue=queue)
+
+            self._save_attributes()
+
+        def _onedal_predict_supported(self, method_name, X, sample_weight=None):
+            assert method_name == "predict"
```

Comment:
When score gets implemented, it will probably have to resuse the _onedal_predict_supported (this code looks to match the style of the LinearRegression implementation, so some adaptation for score has already been done and can be referred to).